tcl/image_filesystem_cache.tcl

```diff
@@ -0,0 +1,339 @@
+}
+
+proc qc::image_filesystem_cache_data {args} {
+    #| Return filesystem cache data
+    #| args: dict-style of cache_dir, file_id, width, height, autocrop
+    qc::args2vars $args {*}{
+        cache_dir
+        file_id
+        max_width
+        max_height
+        autocrop
+    }
+
+    # Requested size matches or exceeds full-sized image in cache
+    if { [qc::image_filesystem_cache_smaller_biggest_exists {*}$args] } {
+        if { $autocrop } {
+            set data [qc::image_filesystem_cache_autocrop_data \
+                        ~ file_id cache_dir]
+            qc::image_nsv_cache_autocrop_set $file_id data
+
+        } else {
+            set data [qc::image_filesystem_cache_original_data \
+                        ~ file_id cache_dir]
+            qc::image_nsv_cache_original_set $file_id data
+        }
+        return $data
+    }
+
+    foreach file [qc::image_filesystem_cache_glob {*}$args] {
+        lassign \
+            [qc::image_filesystem_cache_file2dimensions $file] \
+            width height
+
+        if { ( $width == $max_width
+               &&
+               $height <= $max_height )
+             ||
+             ( $height == $max_height
+               &&
+               $width <= $max_width )
+         } {
+            set url [qc::file2url $file]
+            set timestamp [qc::cast timestamp [file mtime $file]]
+            set data [dict_from width height url timestamp]
+
+            qc::image_nsv_cache_set ~ autocrop file_id data
+            return $data
+        }
+    }
+
+    error "No matching filesystem image data cache exists"
+}
+
+proc qc::image_filesystem_cache_create {args} {
+    #| Create a file for this image in the disk cache,
+    #| constrained to max_width & max_height,
+    #| optionally auto-cropped,
+    qc::args $args -autocrop -- cache_dir file_id max_width max_height
+    default autocrop false
+
+    if { $autocrop } {
+        dict2vars [qc::image_cache_autocrop_data ~ cache_dir file_id] \
+            file width height
+    } else {
+        dict2vars [qc::image_cache_original_data ~ cache_dir file_id] \
+            file width height
+    }
+    set ext [file extension $file]
+
+    if { $width > $max_width
+         || $height > $max_height
+     } {
+        set file [qc::image_file_resize $file $max_width $max_height]
+        dict2vars [qc::image_file_info $file] width height
+
+        set file_path_parts [list $cache_dir]
+        if { $autocrop } {
+            lappend file_path_parts \
+                $file_id \
+                autocrop \
+                ${width}x${height}
+        } else {
+            lappend file_path_parts \
+                ${file_id}-${width}x${height}
+        }
+        lappend file_path_parts ${file_id}${ext}
+
+        set cache_file [join $file_path_parts "/"]
+
+        file mkdir [file dirname $cache_file]
+        file rename -force $file $cache_file
+    }
+}
+
+proc qc::image_filesystem_cache_smaller_biggest_exists {args} {
```

Comment:
Not sure... it's "If the biggest (original) version of the image exists, and is smaller than (or equal to) the specified constraints".
`qc::image_filesystem_cache_smaller_original_exists`?
`qc::image_filesystem_cache_biggest_exists_and_is_smaller`?